### PR TITLE
[extractor/Facebook] fix: Changed json data retrieval

### LINKED ISSUE #7901: Facebook: Cannot parse data (ALL links affected)
### Checklist

- [X] I'm reporting that yt-dlp is broken on a **supported** site
- [X] I've verified that I'm running yt-dlp version **2023.07.06** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/yt-dlp/yt-dlp/wiki/FAQ#video-url-contains-an-ampersand--and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

Slovakia

### Provide a description that is worded well enough to be understood

I tried several fb video urls, all with the same result `Cannot parse data; please report this issue` Worked fine not that long ago.

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [ ] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
yt-dlp.exe -vU https://www.facebook.com/100044498826847/videos/31920791
0501223
[debug] Command-line config: ['-vU', 'https://www.facebook.com/100044498826847/v
ideos/319207910501223']
[debug] Encodings: locale cp1250, fs utf-8, pref cp1250, out utf-8 (No VT), erro
r utf-8 (No VT), screen utf-8 (No VT)
[debug] yt-dlp version stable@2023.07.06 [b532a3481] (win_exe)
[debug] Python 3.8.10 (CPython AMD64 64bit) - Windows-7-6.1.7601-SP1 (OpenSSL 1.
1.1k  25 Mar 2021)
[debug] exe versions: ffmpeg 6.0-full_build-www.gyan.dev (setts), ffprobe 6.0-fu
ll_build-www.gyan.dev
[debug] Optional libraries: Cryptodome-3.18.0, brotli-1.0.9, certifi-2023.05.07,
 mutagen-1.46.0, sqlite3-2.6.0, websockets-11.0.3
[debug] Proxy map: {}
[debug] Loaded 1855 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releas
es/latest
Available version: stable@2023.07.06, Current version: stable@2023.07.06
Current Build Hash: 5ff3e702171a50175c34397494e2d18ce35d771c2110b1e59bd173ec2fb3
52aa
yt-dlp is up to date (stable@2023.07.06)
[facebook] Extracting URL: https://www.facebook.com/100044498826847/videos/31920
7910501223
[facebook] 319207910501223: Downloading webpage
[facebook] 319207910501223: Downloading webpage
ERROR: [facebook] 319207910501223: Cannot parse data; please report this issue o
n  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issu
e template. Confirm you are on the latest version using  yt-dlp -U
  File "yt_dlp\extractor\common.py", line 710, in extract
  File "yt_dlp\extractor\facebook.py", line 707, in _real_extract
  File "yt_dlp\extractor\facebook.py", line 654, in _extract_from_url
```

### PR DESCRIPTION
**Pull request by ringus1**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

Facebook looks to have changed API and json data is served bit differently.

Fixes #7901 

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details><details><summary>Copilot Summary</summary>  

### Summary
🛠️🌟➕

Fix and enhance `FacebookIE` extractor to handle new webpage layout and more video URLs. Update tests, metadata, and formats accordingly.

> _The FacebookIE extractor was broken_
> _By a change that the website had spoken_
> _But this pull request_
> _Fixes it and adds zest_
> _With more URLs and metadata tokens_

### Walkthrough
*  Fix extraction of video formats, subtitles, and thumbnails from Facebook webpages ([link](https://github.com/yt-dlp/yt-dlp/pull/7890/files?diff=unified&w=0#diff-ca6052d2c84eb4ac45aaf04463a4eaecdc75fb3ae4766e234917d8ab131b60ecL404-R424), [link](https://github.com/yt-dlp/yt-dlp/pull/7890/files?diff=unified&w=0#diff-ca6052d2c84eb4ac45aaf04463a4eaecdc75fb3ae4766e234917d8ab131b60ecL496-R514), [link](https://github.com/yt-dlp/yt-dlp/pull/7890/files?diff=unified&w=0#diff-ca6052d2c84eb4ac45aaf04463a4eaecdc75fb3ae4766e234917d8ab131b60ecR522-R526), [link](https://github.com/yt-dlp/yt-dlp/pull/7890/files?diff=unified&w=0#diff-ca6052d2c84eb4ac45aaf04463a4eaecdc75fb3ae4766e234917d8ab131b60ecL514-R537), [link](https://github.com/yt-dlp/yt-dlp/pull/7890/files?diff=unified&w=0#diff-ca6052d2c84eb4ac45aaf04463a4eaecdc75fb3ae4766e234917d8ab131b60ecL529-R553))
* Update uploader_id values in existing test cases for FacebookIE extractor ([link](https://github.com/yt-dlp/yt-dlp/pull/7890/files?diff=unified&w=0#diff-ca6052d2c84eb4ac45aaf04463a4eaecdc75fb3ae4766e234917d8ab131b60ecL100-R118), [link](https://github.com/yt-dlp/yt-dlp/pull/7890/files?diff=unified&w=0#diff-ca6052d2c84eb4ac45aaf04463a4eaecdc75fb3ae4766e234917d8ab131b60ecL182-R200), [link](https://github.com/yt-dlp/yt-dlp/pull/7890/files?diff=unified&w=0#diff-ca6052d2c84eb4ac45aaf04463a4eaecdc75fb3ae4766e234917d8ab131b60ecL277-R295))
* Add a new test case for FacebookIE extractor with a different URL and info_dict ([link](https://github.com/yt-dlp/yt-dlp/pull/7890/files?diff=unified&w=0#diff-ca6052d2c84eb4ac45aaf04463a4eaecdc75fb3ae4766e234917d8ab131b60ecL76-R94))

</details>

### CODE REVIEW

File: yt_dlp/extractor/facebook.py [hunk 1]
```diff
@@ -74,6 +74,22 @@ class FacebookIE(InfoExtractor):
     _VIDEO_PAGE_TAHOE_TEMPLATE = 'https://www.facebook.com/video/tahoe/async/%s/?chain=true&isvideo=true&payloadtype=primary'
 
     _TESTS = [{
+        'url': 'https://www.facebook.com/radiokicksfm/videos/3676516585958356/',
+        'info_dict': {
+            'id': '3676516585958356',
+            'ext': 'mp4',
+            'title': 'dr Adam Przygoda',
+            'description': 'DR ADAM PRZYGODA LIVE (przepraszamy na początku rozmowy nie ma dźwięku ) Bilety na II Konferencja Zdrowia "Łączy Nas dbałość o zdrowie\' Londyn 2023 | Targi Zdrowia   znajdziesz na Www.konferencjalondyn.co.uk \nRADIO KICKS FM  jest głównym patronem medialnym wydarzenia.',
```
Comment (bashonly) [inline yt_dlp/extractor/facebook.py:82]: Use md5 for long descriptions

File: yt_dlp/extractor/facebook.py [hunk 2]
```diff
@@ -401,9 +417,9 @@ def _extract_from_url(self, url, video_id):
 
         def extract_metadata(webpage):
             post_data = [self._parse_json(j, video_id, fatal=False) for j in re.findall(
-                r'handleWithCustomApplyEach\(\s*ScheduledApplyEach\s*,\s*(\{.+?\})\s*\);', webpage)]
+                r'data-sjs>({.*?ScheduledServerJS.*?})<\/script', webpage)]
```
Comment (bashonly) [inline yt_dlp/extractor/facebook.py:420]:
```suggestion
                r'data-sjs>({.*?ScheduledServerJS.*?})</script>', webpage)]
```

File: yt_dlp/extractor/facebook.py [hunk 3]
```diff
@@ -493,14 +509,19 @@ def process_formats(info):
 
         def extract_relay_data(_filter):
             return self._parse_json(self._search_regex(
-                r'handleWithCustomApplyEach\([^,]+,\s*({.*?%s.*?})\);' % _filter,
+                r'data-sjs>({.*?%s.*?})<\/script' % _filter,
```
Comment (bashonly) [inline yt_dlp/extractor/facebook.py:512]:
```suggestion
                r'data-sjs>({.*?%s.*?})</script>' % _filter,
```

File: yt_dlp/extractor/facebook.py [hunk 4]
```diff
@@ -493,14 +509,19 @@ def process_formats(info):
...
         def extract_relay_prefetched_data(_filter):
             replay_data = extract_relay_data(_filter)
             for require in (replay_data.get('require') or []):
                 if require[0] == 'RelayPrefetchedStreamCache':
                     return try_get(require, lambda x: x[3][1]['__bbox']['result']['data'], dict) or {}
+                else:
+                    data = traverse_obj(require, (
+                        ..., ..., '__bbox', 'require', lambda _, v: "RelayPrefetchedStreamCache" in v,
+                        ..., ..., "__bbox", "result", "data"))
+                    return data[0] if data else {}
```
Comment (bashonly) [inline yt_dlp/extractor/facebook.py:524]:
```suggestion
                else:
                    return traverse_obj(require, (
                        ..., ..., '__bbox', 'require', lambda _, v: 'RelayPrefetchedStreamCache' in v,
                        ..., ..., '__bbox', 'result', 'data', {dict}), get_all=False) or {}
```
Do we need the `if/else` though? From testing a bit, it seems like the `if require[0]...` branch might be obsolete now. I could be wrong

File: yt_dlp/extractor/facebook.py [hunk 5]
```diff
@@ -511,7 +532,7 @@ def extract_relay_prefetched_data(_filter):
 
         if not video_data:
             data = extract_relay_prefetched_data(
-                r'"(?:dash_manifest|playable_url(?:_quality_hd)?)"\s*:\s*"[^"]+"')
+                r'"(?:dash_manifest|playable_url(?:_quality_hd)?)')
```
Comment (bashonly) [inline yt_dlp/extractor/facebook.py:535]: Is this necessary? I'm not saying it's wrong to change it, just wondering why

**Review by bashonly — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/facebook.py (hunk 4 above)
Comment (pukkandan) [reply yt_dlp/extractor/facebook.py:524]: facebook has so many edge cases. It's best to not remove old patterns

File: yt_dlp/extractor/facebook.py (hunk 5 above)
Comment (ringus1) [reply yt_dlp/extractor/facebook.py:535]: Keeping that trailing pattern breaks at least 1 test actually: `https://www.facebook.com/video.php?v=274175099429670`
And I don't see a reason to keep it anyway. Dependent pattern in `extract_relay_data` makes sure we collect only valid json body of `script`s. This pattern just narrows it down a bit.

File: yt_dlp/extractor/facebook.py (hunk 4 above)
Comment (ringus1) [reply yt_dlp/extractor/facebook.py:524]: It LOOKS obsolete indeed. Is there any common function that would allow finding:
```
lambda _, v: 'RelayPrefetchedStreamCache' in v, ..., ..., '__bbox', 'result', 'data', {dict}
```
as deep as needed to cover both of these cases in 1?

File: yt_dlp/extractor/facebook.py [hunk 6]
```diff
@@ -493,14 +509,19 @@ def process_formats(info):
...
         def extract_relay_prefetched_data(_filter):
             replay_data = extract_relay_data(_filter)
             for require in (replay_data.get('require') or []):
                 if require[0] == 'RelayPrefetchedStreamCache':
                     return try_get(require, lambda x: x[3][1]['__bbox']['result']['data'], dict) or {}
+                else:
+                    data = traverse_obj(require, (
+                        ..., ..., '__bbox', 'require', lambda _, v: "RelayPrefetchedStreamCache" in v,
+                        ..., ..., "__bbox", "result", "data"))
+                    return data[0] if data else {}
```
Comment (bashonly) [inline yt_dlp/extractor/facebook.py:524]:
```suggestion
        def extract_relay_prefetched_data(_filter):
            return traverse_obj(extract_relay_data(_filter), (
                'require', (None, (..., ..., ..., '__bbox', 'require')),
                lambda _, v: 'RelayPrefetchedStreamCache' in v, ..., ...,
                '__bbox', 'result', 'data', {dict}), get_all=False) or {}
```

File: yt_dlp/extractor/facebook.py (hunk 4 above)
Comment (bashonly) [reply yt_dlp/extractor/facebook.py:524]: > Is there any common function that would allow finding...

https://github.com/yt-dlp/yt-dlp/pull/7890/files#r1299998111

File: yt_dlp/extractor/facebook.py (hunk 6 above)
Comment (ringus1) [reply yt_dlp/extractor/facebook.py:524]: Cool, but that's just a merge of those. Is there some equivalent of xpath `//require[...]`? (For future reference)

Comment (bashonly) [reply yt_dlp/extractor/facebook.py:524]: afaik no

**Review by bashonly — APPROVED** [review]

**Comment by kalemi19** [comment]
Any ETA for when this will be merged into the Master branch / new release?

File: yt_dlp/extractor/facebook.py [hunk 7]
```diff
@@ -401,9 +417,9 @@ def _extract_from_url(self, url, video_id):
 
         def extract_metadata(webpage):
             post_data = [self._parse_json(j, video_id, fatal=False) for j in re.findall(
-                r'handleWithCustomApplyEach\(\s*ScheduledApplyEach\s*,\s*(\{.+?\})\s*\);', webpage)]
+                r'data-sjs>({.*?ScheduledServerJS.*?})</script>', webpage)]
```
Comment (pukkandan) [inline yt_dlp/extractor/facebook.py:420]: This can cause catastrophic backtracking. Try to see if there's a better way. Can still merge if no easy solution is found, but it is something to be aware of.

File: yt_dlp/extractor/facebook.py [hunk 8]
```diff
@@ -493,14 +509,14 @@ def process_formats(info):
 
         def extract_relay_data(_filter):
             return self._parse_json(self._search_regex(
-                r'handleWithCustomApplyEach\([^,]+,\s*({.*?%s.*?})\);' % _filter,
+                r'data-sjs>({.*?%s.*?})</script>' % _filter,
```
Comment (pukkandan) [inline yt_dlp/extractor/facebook.py:512]: same

**Review by pukkandan — APPROVED** [review]

File: yt_dlp/extractor/facebook.py (hunk 7 above)
Comment (ringus1) [reply yt_dlp/extractor/facebook.py:420]: Well, this is potential replacement:
```
r'data-sjs>([^<]+)</script>', webpage)]
```
Just traverse in the next line will have more items to browse through. Not sure if this would provide better performance since `webpage` data that is collected is up to few MBs in size. Not GBs or something.

File: yt_dlp/extractor/facebook.py (hunk 8 above)
Comment (ringus1) [reply yt_dlp/extractor/facebook.py:512]: This one is more tricky. There is no obvious negative pattern that could optimise it since `_filter` is unknown too. So potential alternative would be to grab all jsons with `r'data-sjs>([^<]+)</script>'` regex, then find the first one matching `_filter`. Again, not sure if this worth considering for such small files.

File: yt_dlp/extractor/facebook.py (hunk 7 above)
Comment (Cyp) [reply yt_dlp/extractor/facebook.py:420]: Haven't tested, but I think a relatively easy way would be `r'data-sjs>({(?:(?!ScheduledServerJS).)*?ScheduledServerJS.*?})</script>'`. A cleaner way would be `r'data-sjs>({(?>.*?ScheduledServerJS).*?})</script>'`, but using atomic groups `(?>…)` would unfortunately require Python 3.11.

And similar for the `%s` below.

P.S. Was there something wrong with my other comment in this thread? It says it was hidden.

Comment (ringus1) [reply yt_dlp/extractor/facebook.py:420]: @Cyp I'm not too much of an expert in regex optimisation, but it looks to be more complicated by testing it in https://regex101.com/

Comment (bashonly) [reply yt_dlp/extractor/facebook.py:420]: > P.S. Was there something wrong with my other comment in this thread? It says it was hidden.

I hid the comment because bb5d84c9d2f1e978c3eddfb5ccbe138036682a36 already does what you suggested (you must not be working from master)

Comment (Cyp) [reply yt_dlp/extractor/facebook.py:420]: I was working with a distro version, with some patch files I semi-auto-applied from 7564 and 7890. Looks like one important chunk of the 7564.patch file silently failed to be applied even though other parts were (I can see it in the patch file, but after applying the patch, that part is ignored somehow). Anyway, thanks, good to know it's fixed in master then.

**Comment by bashonly** [comment]
I am noticing that FB now serves different formats/codecs depending on the user-agent header.

e.g. with yt-dlp's default chrome UA headers, when we extract `https://www.facebook.com/reel/1481547595669565` (from someone else's earlier comment), we'll get only vp9 and av1 vcodecs. 

But if we pass `--user-agent "Mozilla/5.0 (Windows NT 10.0; rv:109.0) Gecko/20100101 Firefox/117.0"` (FF UA), we can get the avc vcodec formats of equivalent resolutions.

I'm wondering if we should take the `dash_manifest_url`, for which extraction would usually result in duplicate formats, and extract it with a FF UA instead to get all possible formats

Something like this (very rough draft):
```diff
diff --git a/yt_dlp/extractor/facebook.py b/yt_dlp/extractor/facebook.py
index c30a6b06a..53498acb8 100644
--- a/yt_dlp/extractor/facebook.py
+++ b/yt_dlp/extractor/facebook.py
@@ -543,12 +542,15 @@ def parse_graphql_video(video):
                     q = qualities(['sd', 'hd'])
                     for key, format_id in (('playable_url', 'sd'), ('playable_url_quality_hd', 'hd'),
                                            ('playable_url_dash', ''), ('browser_native_hd_url', 'hd'),
-                                           ('browser_native_sd_url', 'sd')):
+                                           ('browser_native_sd_url', 'sd'), ('dash_manifest_url', '')):
                         playable_url = video.get(key)
                         if not playable_url:
                             continue
                         if determine_ext(playable_url) == 'mpd':
-                            formats.extend(self._extract_mpd_formats(playable_url, video_id))
+                            formats.extend(self._extract_mpd_formats(
+                                playable_url, video_id, headers={
+                                    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; rv:109.0) Gecko/20100101 Firefox/117.0',
+                                }))
                         else:
                             formats.append({
                                 'format_id': format_id,
```

**Comment by gamer191** [comment]
> But if we pass `--user-agent "Mozilla/5.0 (Windows NT 10.0; rv:109.0) Gecko/20100101 Firefox/117.0"` (FF UA), we can get the avc vcodec formats of equivalent resolutions.

what happens if you pass user-agents for newer versions of Chrome?

**Comment by bashonly** [comment]
any chrome UA yields vp9/av1

**Comment by ringus1** [comment]
@bashonly Hm, though indeed it seems to act as you say, I see this comment few lines back:
```
            # Downloads with browser's User-Agent are rate limited. Working around
            # with non-browser User-Agent.
```
So, I believe this should be somehow configurable maybe by user in options? I wouldn't stretch and use it by default though because of the above.

And I guess it could be part of another MR?

PS. @pukkandan @bashonly Do you guys have any preferences on the regex final shape given my comments?

**Comment by bashonly** [comment]
> I see this comment few lines back

Pretty sure that comment only applies to format URLs, since the code was previously requesting MPD manifests with the default/chrome UA instead of the `facebookexternalhit` UA

> So, I believe this should be somehow configurable maybe by user in options

Perhaps yeah

> And I guess it could be part of another PR

Yeah sure

> Do you guys have any preferences on the regex final shape given my comments?

The FB webpage html/JS is such an ugly monstrosity that I'm having a hard time finding a better way. It looks like one of the previous patterns was already risking catastrophic backtracking, which we never got any reports of, so I'm not that concerned about it. I'll give it another look a bit later today (but I will probably give up and end up merging as-is)
